max30102.c

```diff
@@ -0,0 +1,222 @@
+    }
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t MAX30102_ReadFIFO(I2C_HandleTypeDef *hi2c, uint32_t *red_data, uint32_t *ir_data) {
+    uint8_t fifo_data[6]; 
+
+    // Read 6 bytes from the FIFO data register
+    if (HAL_I2C_Mem_Read(hi2c,
+                         MAX30102_I2C_ADDRESS,
+                         MAX30102_REG_FIFO_DATA,
+                         I2C_MEMADD_SIZE_8BIT,
+                         fifo_data,
+                         6,
+                         HAL_MAX_DELAY) != HAL_OK)
+    {
+        return MAX30102_ERROR;
+    }
+    
+    // Intensity of red LED reflected back into MAX30102 photodetector
+    *red_data = ((uint32_t)fifo_data[0] << 16) | ((uint32_t)fifo_data[1] << 8) | fifo_data[2];
+    // Intensity of infrared LED reflected back into MAX30102 photodetector
+    *ir_data = ((uint32_t)fifo_data[3] << 16) | ((uint32_t)fifo_data[4] << 8) | fifo_data[5]; 
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t ON(I2C_HandleTypeDef *hi2c) {
+    uint8_t data;
+
+    // Read current value of the SHDN bit
+    if (MAX30102_ReadRegister(hi2c, MAX30102_REG_MODE_CONFIG, &data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+    data &= ~0x80; // Wake up sensor by clearing SHDN bit
+    if (MAX30102_WriteRegister(hi2c, MAX30102_REG_MODE_CONFIG, data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t OFF(I2C_HandleTypeDef *hi2c) {
+    uint8_t data;
+
+    // Read current value of the SHDN bit
+    if (MAX30102_ReadRegister(hi2c, MAX30102_REG_MODE_CONFIG, &data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+    data |= 0x80; // Put sensor to sleep using SHDN bit
+    if (MAX30102_WriteRegister(hi2c, MAX30102_REG_MODE_CONFIG, data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+
+
+    return MAX30102_OK;
+}
+/*HOW DOES HEART RATE SENSOR WORK
+
+Red and infrared data signals are read by the MAX30102 sensor. These values are a result of the reflected light intensity after the red and infrared LEDs hit a blood vessel
+The reflection intensity will be low during a heart beat since blood volume is higher in the arteries. Higher blood volume leads to slightly darker skin around the finger thus more light
+is absorbed. Since more light is absorbed, less will be reflected. Thus in graphing terms, a trough is reached when the heart beats.
+The opposite is true between heart beats, blood volume is lower leading to a lighter color which absorbs less light and reflects more. In turn a peak is reached.
+Calculating the time between peaks can allow an estimated heart rate to be calculated.
+
+Infrared data is primarily used for heart rate, data sheet states red data is 'critical' for SPo2 which is necessary for temperature sensing
+*/
+
+MAX30102_Status_t GetHeartRate(I2C_HandleTypeDef *hi2c, float *heart_rate, int num_peaks) { // Number of peaks before heart rate calculation necessary on function call
```

Comment:
What does it do?: 
Red and infrared data signals are read by the MAX30102 sensor. These values are a result of the reflected light intensity after the red and infrared LEDs hit a blood vessel The reflection intensity will be low during a heart beat since blood volume is higher in the arteries. Higher blood volume leads to slightly darker skin around the finger thus more light is absorbed. Since more light is absorbed, less will be reflected. Thus in graphing terms, a trough is reached when the heart beats. The opposite is true between heart beats, blood volume is lower leading to a lighter color which absorbs less light and reflects more. In turn a peak is reached. Calculating the time between peaks can allow an estimated heart rate to be calculated. Infrared data is primarily used for heart rate, data sheet states red data is 'critical' for SPo2 which is necessary for temperature sensing.